Global conditions

Toilal opened the pull request:
This is a big refactoring of conditions used by await matchers. It's based on #265, but with more breaking changes. It introduce interfaces for conditions used in await().until() matchers.

Those interfaces are implemented by Await matchers, but also by a conditions() object available in FluentWebElement and FluentList. So basically, conditions are now available without having to perform a await().until() call.

It makes codebase more maintainable and readable, and also supports negation of condition easily (using not() function). There's still a bit of work to include latest conditions introduced in the project, like ```hasPositionX``` and ```isAboveScreenOrInvisible```.

I would like to write more test on those conditions too. Currently all previous tests have been ported and runs properly, but it could probably be enhanced. Maybe it will come later.

Feel free to review it if you have time, but I would like to merge quickly now (tomorrow).

filipcynarski commented on fluentlenium-core/src/main/java/org/fluentlenium/core/wait/FluentWaitElementListMatcher.java:
```diff
@@ -23,6 +24,13 @@ public FluentWaitElementListMatcher(Search search, FluentWait fluentWait, List<?
 
     @Override
+    public FluentWaitElementListMatcher not() {
+        FluentWaitElementListMatcher ret = new FluentWaitElementListMatcher(search, wait, untilElements);
```
ret name doesn't make sense to me, can you set something meaningful?

filipcynarski commented on fluentlenium-core/src/main/java/org/fluentlenium/core/wait/FluentWaitIntegerMatcher.java:
```diff
@@ -0,0 +1,110 @@
+    private final AbstractWaitElementListMatcher matcher;
+
+    public FluentWaitIntegerMatcher(AbstractWaitElementListMatcher matcher) {
+        this.matcher = matcher;
```
why we are using this?

Toilal replied:
In this case, it's mandatory to use this because local variable and class variable have the same name (matcher).

filipcynarski commented on fluentlenium-core/src/main/java/org/fluentlenium/core/wait/FluentWaitElementListMatcher.java:
```diff
@@ -23,6 +24,13 @@ public FluentWaitElementListMatcher(Search search, FluentWait fluentWait, List<?
+    @Override
     protected FluentList<FluentWebElement> find() {
         FluentListImpl<FluentWebElement> elements = new FluentListImpl<>();
         if (this.untilElements != null) {
```
is `this` operator required here

Toilal replied:
In the case, it should be removed :) I'll force push the latest commit

filipcynarski commented:
Generally I think it is really good change. Thank you for working on that. What I'm thinking it is bumping release number to 1.0.0 because it is breaking change and we need describe it in details. People who are using our project should spend a lot of time adjusting theirs automation to these changes. What do you think? We can release the snapshot first, test it on our projects, improve test coverage and all pending tasks and if we successfully manage that we can perform the major release then.

filipcynarski commented on fluentlenium-core/src/test/java/org/fluentlenium/integration/FluentLeniumWaitTest.java:
```diff
@@ -152,37 +182,37 @@ public void checkAwaitContainsTextWithTextMatcher() {
     @Test
     public void when_a_element_is_not_present_then_isNotPresent_return_true() {
-        await().atMost(1, NANOSECONDS).until(".small").withText().contains("notPresent").isNotPresent();
+        await().atMost(1, NANOSECONDS).until(".small").withText().contains("notPresent").not().isPresent();
```
I'm not sure `isPresent` is still good mayeb it would be better to replace it with just `present` ? It would be easier to read. Not is present reads strange to me.

Toilal replied:
Problem is that this method name is in an interface, and it would seems strange in other use case to have present() only. Conditions are all prefixed with `is` or `has`, I think we should stick with this consistently.

filipcynarski commented on fluentlenium-core/src/test/java/org/fluentlenium/integration/FluentLeniumWaitTest.java:
```diff
@@ -283,49 +313,49 @@ public void when_element_is_not_displayed_then_isDisplayed_throws_exception() {
     public void when_element_is_not_present_then_areDisplayed_throws_exception() {
         goTo(JAVASCRIPT_URL);
-        await().atMost(1, NANOSECONDS).until("#nonexistent").areDisplayed();
+        await().atMost(1, NANOSECONDS).until("#nonexistent").each().isDisplayed();
```
the same as in `isPresent` case

filipcynarski commented on fluentlenium-core/src/test/java/org/fluentlenium/integration/FluentLeniumWaitTest.java:
```diff
@@ -283,49 +313,49 @@ public void when_element_is_not_displayed_then_isDisplayed_throws_exception() {
     public void when_element_is_not_present_then_areEnabled_throws_exception() {
         goTo(JAVASCRIPT_URL);
-        await().atMost(1, NANOSECONDS).until("#nonexistent").areEnabled();
+        await().atMost(1, NANOSECONDS).until("#nonexistent").each().isEnabled();
```
the same as in `isPresent` case

Toilal commented:
No problem for going to 1.0.0. We could release a beta version first.

I think i'll open a `1.x` branch then. After final release, `1.x` will become master, and master branch will become `0.x` branch for maintenance purpose if required.

Toilal commented:
Also, i'll update README docs accordingly.

Toilal commented:
In fact, master will be for `1.x` and i'll create a 0.x branch for maintenance branch of 0 version.

Toilal commented:
I've introduced Rectangle conditions that supports any condition on Rectangle of an element. So I have removed function hasPositionX() as it's replaced by hasRectangle().withX()).

I've also marked isAboveScreenOrInvisible as deprecated, as I think it's too complex and project specific. Users can use is ```isVerified(Predicate)``` function instead to implement their custom condition without modifying FluentLenium sources.

I plan to rewrite from scratch complete tests for conditions() and await() feature in the afternoon.
